### PR TITLE
[AutoDiff] Remove 'differentiableFunction(from:)' and 'linearFunction(from:)'

### PR DESCRIPTION
**Pull request by rxwei**
Remove unused APIs `differentiableFunction(from:)` and `linearFunction(from:)`. They were never official APIs, are not included in the [initial proposal](https://github.com/rxwei/swift-evolution/blob/autodiff/proposals/0000-differentiable-programming.md#make-a-function-differentiable-using-derivative), and are unused by existing supported client libraries (SwiftFusion and S4TF). Most importantly, they block crucial optimizations on linear map closures (#34935) and would need nontrivial work in SILGen to support.

### CODE REVIEW

**Comment by rxwei** [comment]
@swift-ci please test

**Review by dan-zheng — APPROVED** [review]
LGTM! I'm not familiar with existing use cases or reasons not to remove - my memory is that you added them as a cute idea.

**Comment by rxwei** [comment]
@swift-ci please test

**Comment by porterchild** [comment]
@rxwei @dan-zheng  I was making use of `differentiableFunction` for checkpointing, is there a way to achieve gradient checkpointing without `differentiableFunction`? Like [here](https://www.tensorflow.org/swift/tutorials/custom_differentiation#recomputing_activations_during_backpropagation_to_save_memory_checkpointing)

**Comment by porterchild** [comment]
Would something like this do it?
```
@differentiable
func makeRecomputedInGradient<T: Differentiable, U: Differentiable>(
    _ original: @escaping @differentiable(T) -> U
) -> @differentiable(T) -> U {
    
    @differentiable
    func newFunc(_ t: T) -> U {
        var result: U? = nil
        var result = result.withDerivative { v in v * pullback(at: t, in: original) }
        result = withoutDerivative(at: original(t))
        return result
    }
    
    return newFunc
}
```

**Comment by rxwei** [comment]
```swift
func _original<T, R>(_ x: T, _ f: @differentiable (T) -> R) -> R {
  f(x)
}

@derivative(of: _original, wrt: x)
func _vjp<T, R>(
  _ x: T, _ f: @differentiable (T) -> R
) -> (value: R, pullback: (R.TangentVector) -> T.TangentVector) {
  (value: f(x), pullback: pullback(at: x, in: f))
}

func makeRecomputed<T, R>(
  _ f: @escaping @differentiable (T) -> R
) -> @differentiable (T) -> R {
  { x in _original(x, f) }
}
```

**Comment by porterchild** [comment]
Thanks Richard!
